[RF-DOCS] Wishlists guide (#55428)
rails/rails: guides/source/wishlists.md

**DO NOT READ THIS FILE ON GITHUB, GUIDES ARE PUBLISHED ON <https://guides.rubyonrails.org>.**

Wishlists
=========

This guide covers adding Wishlists to the store e-commerce application in the
[Getting Started Guide](getting_started.html)). We will use the code from the

> [Sign up and Settings Guide](sign_up_and_settings.html) as a starting place.

Reviewer: This came up in conversation with @excid3 when we were doing the first one, but the longterm maintenance as @MatheusRich says would be an issue. 
@rafaelfranca Could this be in the Rails GitHub or would do you think an external repo is better?

After reading this guide, you will know how to:

* Add wishlists
* Use counter caches
* Add friendly URLs
* Filter records

--------------------------------------------------------------------------------

Introduction
------------

E-commerce stores often have wishlists for sharing products. Customers can use
wishlists to keep track of products they'd like to buy or share them with
friends and family for gift ideas.

Let's get started!

Wishlist Models
---------------

Our e-commerce store has Products and Users that we built in the previous
tutorials. These are the foundations we need to build Wishlists. Each wishlist
belongs to a user and contains a list of Products.

Let's start by creating the Wishlist model.

```bash
$ bin/rails generate model Wishlist user:belongs_to name products_count:integer
```

This model has 3 attributes:

- `user:belongs_to` which associates the Wishlist with the `User` record who
  owns it
- `name` which we'll also use for friendly URLs

> - `products_count` for the counter cache we'll add to count how many products

Reviewer 2: I think we never introduce counter caches (maybe in previous tutorials?). Anyway, a link to the docs wouldn't hurt.
```suggestion
- `products_count` for the [counter cache](https://guides.rubyonrails.org/association_basics.html#counter-cache) we'll add to count how many products
```

access the `Wishlists` and Rails will know to automatically `JOIN` the tables in
SQL queries.

We also set `wishlist_products` as `dependent: :destroy`. When a `Product` is
destroyed, it will be automatically removed from any Wishlists.

In `app/models/wishlist.rb`, let's update both associations to enable counter
caching:

```ruby#2-5
class WishlistProduct < ApplicationRecord
  belongs_to :product, counter_cache: :wishlists_count
  belongs_to :wishlist, counter_cache: :products_count

  validates :product_id, uniqueness: { scope: :wishlist_id }
end
```

We've specified a column name to update on the associated models. For the
Product model, we want to use the `wishlists_count` column and for `Wishlist` we
want to use `products_count`. These counter caches update anytime a
`WishlistProduct` is created or destroyed.

The uniqueness validation also tells Rails to check if a product is already on
the wishlist. This is paired with the unique index on the wishlist_product table
so that it's also validated at the database level.

Finally, let's update `app/models/wishlist.rb` with it's associations:

```ruby
class Wishlist < ApplicationRecord
  belongs_to :user

>   has_many :wishlist_products, dependent: :destroy

Reviewer 2: Could be, if you think of this as an adjective ("products that were wishlisted"), but wishlist_products also work as "products that are on a wishlist".

  has_many :products, through: :wishlist_products
end
```

Just like with `Product`, `wishlist_products` uses the `dependent: :destroy`
option to automatically remove join table records when a Wishlist is deleted.

### Friendly URLs

Wishlists are often shared with friends and family. By default, the ID in the
URL for a `Wishlist` is a simple Integer. This means we can't easily look at the
URL to determine which `Wishlist` it's for.

Active Record uses `to_param` method that can be used for generating pretty
URLs. Let's try it out by creating the `to_param` method in our model:

```ruby#6-8
class Wishlist < ApplicationRecord
  belongs_to :user
  has_many :wishlist_products, dependent: :destroy
  has_many :products, through: :wishlist_products

  def to_param
    "#{id}-#{name.squish.parameterize}"
  end
end
```

> This returns a String for the URL param made up of the `id` and `name` joined by
> hyphens. `name` is made URL safe by using
> [`squish`](https://api.rubyonrails.org/classes/String.html#method-i-squish) to
> clean up whitespace and
> [`parameterize`](https://api.rubyonrails.org/classes/String.html#method-i-parameterize)
> to replace special characters.

Reviewer 2: ```suggestion
This will create a `to_param`  instance method that returns a String for the URL param made up of the `id` and `name` joined by
hyphens. `name` is made URL safe by using
[`squish`](https://api.rubyonrails.org/classes/String.html#method-i-squish) to
clean up whitespace and
[`parameterize`](https://api.rubyonrails.org/classes/String.html#method-i-parameterize)
to replace special characters.
```

  def create
    @wishlist.wishlist_products.create(product: @product)
    redirect_to @wishlist, notice: "#{@product.name} added to wishlist."
  end

  private
    def set_product
      @product = Product.find(params[:product_id])
    end

    def set_wishlist
      @wishlist = Current.user.wishlists.find(params[:wishlist_id])
    end
end
```

Since we're in a nested resource route, we find the `Product` using the
`:product_id` param.

The `create` action is also simpler than normal. If a product is already on the
wishlist, the `wishlist_product` record will fail to create but we don't need to
notify the user of this error so we can redirect to the wishlist in either case.

Log in as the user we created a Wishlist for earlier and add a product to the
wishlist.

### Default Wishlist

This works fine since we created a `Wishlist` in the Rails console, but what
happens when the user doesn't have any wishlists?

Run the following to delete all wishlists in the database:

```bash
$ bin/rails runner "Wishlist.destroy_all"
```

Try visiting a product and adding it to a Wishlist now.

The first problem is the select box will be empty. The form will not submit a

> `wishlist_id` param to the server and that will cause ActiveRecord to raise an

Reviewer 3: ```suggestion
`wishlist_id` param to the server and that will cause Active Record to raise an
```

  end

  def wishlist_params
    params.expect(wishlist: [ :name ])
  end
end
```

This is a very standard controller with a couple important changes:

- Actions are scoped to `Current.user.wishlists` so only the owner can create,
  update, and delete their own wishlists
- `show` is publicly accessible so wishlists can be shared and viewed by anyone

### Wishlist Views

Create the index view at `app/views/wishlists/index.html.erb`:

```erb
<h1>Your Wishlists</h1>
<%= link_to "Create a wishlist", new_wishlist_path %>
<%= render @wishlists %>
```

This renders the `_wishlist` partial so let's create that at
`app/views/wishlists/_wishlist.html.erb`:

```erb
<div>
  <%= link_to wishlist.name, wishlist %>
</div>
```

Next let's create the `new` view at `app/views/wishlists/new.html.erb`:

```erb
<h1>New Wishlist</h1>
<%= render "form", locals: {wishlist: @wishlist} %>
```

And the `edit` view at `app/views/wishlists/edit.html.erb`:

```erb
<h1>Edit Wishlist</h1>
<%= render "form", locals: {wishlist: @wishlist} %>
```

Along with the `_form` partial at `app/views/wishlists/_form.html.erb`:

```erb
<%= form_with model: @wishlist do |form| %>
  <% if form.object.errors.any? %>

>     <div><%= form.object.errors.full_messages.first %></div>

Reviewer 2: I'm curious about first here. Is that how we've been doing on previous tutorials?

I usually would do .to_sentence but I guess you wan to show one error at a time?

Author: I can't remember, but I agree we should use `to_sentence` here.

```suggestion
    <div><%= form.object.errors.full_messages.to_sentence %></div>
```

Since Rails ships with Hotwire on the frontend, we can use its
[Stimulus framework](https://stimulus.hotwired.dev/) to add some JavaScript
functionality to our UI.

First, let's add a button to `app/views/wishlists/show.html.erb`:

```erb#7
<h1><%= @wishlist.name %></h1>
<% if authenticated? && @wishlist.user == Current.user %>
  <%= link_to "Edit", edit_wishlist_path(@wishlist) %>
  <%= button_to "Delete", @wishlist, method: :delete, data: {turbo_confirm: "Are you sure?"} %>
<% end %

<%= tag.button "Copy to clipboard", data: { controller: :clipboard, action: "clipboard#copy", clipboard_text_value: wishlist_url(@wishlist) } %>
```

This button has several data attributes that wire up to the JavaScript. We're
using the Rails `tag` helper to make this shorter which outputs the following
HTML:

```html
<button data-controller="clipboard" data-action="clipboard#copy" data-clipboard-text-value="/wishlists/1-example-wishlist">
  Copy to clipboard
</button>
```

What do these data attributes do? Let's break down each one:

- `data-controller` tells Stimulus to connect to `clipboard_controller.js`
- `data-action` tells Stimulus when the button is clicked to call the

>   `clipboard` controller's `copy()` function

Reviewer 2: Not to be pedantic, but
```suggestion
  `clipboard` controller's `copy()` method
```

    <% end %>
  </div>
<% end %>
```

Create `app/controllers/wishlists/wishlist_products_controller.rb` with the
following:

```ruby
class Wishlists::WishlistProductsController < ApplicationController
  before_action :set_wishlist
  before_action :set_wishlist_product

  def destroy
    @wishlist_product.destroy
    redirect_to @wishlist, notice: "#{@wishlist_product.product.name} removed from wishlist."
  end

  private

  def set_wishlist
    @wishlist = Current.user.wishlists.find_by(id: params[:wishlist_id])
  end

  def set_wishlist_product
    @wishlist_product = @wishlist.wishlist_products.find(params[:id])
  end
end
```

You can now remove products from any wishlist. Try it out!

### Moving Products To Another Wishlist

With multiple wishlists, users may want to move a product from one list to
another. For example, they might want to move items into a "Christmas" wishlist.

In `app/views/wishlists/show.html.erb`, add the following:

```erb#14-19
<h1><%= @wishlist.name %></h1>
<% if authenticated? && @wishlist.user == Current.user %>
  <%= link_to "Edit", edit_wishlist_path(@wishlist) %>

>   <%= button_to "Delete", @wishlist, method: :delete, data: {turbo_confirm: "Are you sure?"} %>

Reviewer 3: ```suggestion
  <%= button_to "Delete", @wishlist, method: :delete, data: { turbo_confirm: "Are you sure?" } %>
```

<% end %>

<h3><%= pluralize @wishlist.products_count, "Product" %></h3>
<% @wishlist.wishlist_products.includes(:product).each do %>
  <div>
    <%= link_to it.product.name, it.product %>
    <small>Added <%= l it.created_at, format: :long %></small>

    <% if authenticated? && @wishlist.user == Current.user %>
      <% if (other_wishlists = Current.user.wishlists.excluding(@wishlist)) && other_wishlists.any? %>

>         <%= form_with url: [@wishlist, it], method: :patch do |form| %>

Reviewer 3: ```suggestion
        <%= form_with url: [ @wishlist, it ], method: :patch do |form| %>
```

          <%= form.collection_select :new_wishlist_id, other_wishlists, :id, :name %>
          <%= form.submit "Move" %>
        <% end %>
      <% end %>

>       <%= button_to "Remove", [@wishlist, it], method: :delete, data: {turbo_confirm: "Are you sure?"} %>

Reviewer 3: ```suggestion
      <%= button_to "Remove", [ @wishlist, it ], method: :delete, data: { turbo_confirm: "Are you sure?" } %>
```

```erb
<h1><%= @wishlist.name %></h1>
<p>By <%= link_to @wishlist.user.full_name, store_user_path(@wishlist.user) %></p>

<h3><%= pluralize @wishlist.products_count, "Product" %></h3>
<% @wishlist.wishlist_products.includes(:product).each do %>
  <div>
    <%= link_to it.product.name, store_product_path(it.product) %>
    <small>Added <%= l it.created_at, format: :long %></small>
  </div>
<% end %>
```

Lastly, add the link to the sidebar layout:

```erb#14
<%= content_for :content do %>
  <section class="settings">
    <nav>
      <h4>Account Settings</h4>
      <%= link_to "Profile", settings_profile_path %>
      <%= link_to "Email", settings_email_path %>
      <%= link_to "Password", settings_password_path %>
      <%= link_to "Account", settings_user_path %>

      <% if Current.user.admin? %>
        <h4>Store Settings</h4>
        <%= link_to "Products", store_products_path %>
        <%= link_to "Users", store_users_path %>
        <%= link_to "Wishlists", store_products_path %>
      <% end %>
    </nav>

    <div>
      <%= yield %>
    </div>
  </section>
<% end %>

<%= render template: "layouts/application" %>
```

> Now we can view wishlists in the admin area

Reviewer 3: ```suggestion
Now we can view wishlists in the admin area.
```

### Filtering Wishlists

To get a better look at data in the admin area, it's helpful to have filters. We
can filter wishlists by `User` or by `Product`.

Update `app/views/store/wishlists/index.html.erb` by adding the following form:

```erb#1,3-7
<h1><%= pluralize @wishlists.count, "Wishlist" %></h1>

<%= form_with url: store_wishlists_path, method: :get do |form| %>
  <%= form.collection_select :user_id, User.all, :id, :full_name, selected: params[:user_id], include_blank: "All Users" %>
  <%= form.collection_select :product_id, Product.all, :id, :name, selected: params[:product_id], include_blank: "All Products" %>
  <%= form.submit "Filter" %>
<% end %>

<%= render @wishlists %>
```

We've updated the header to display a count of wishlists which will be helpful
when we apply a filter. The form also reads from the URL's query params to
pre-select an option in the select boxes if a filter is applied. This form
submits to the `index` action so it can be used for displaying all records or a
filtered set of records.

To make this work, we need to apply these filters in our SQL query with

> ActiveRecord. Update the controller to include these filters:

Reviewer 3: ```suggestion
Active Record. Update the controller to include these filters:
```

```ruby#4-5
class Store::WishlistsController < Store::BaseController
  def index
    @wishlists = Wishlist.includes(:user)
    @wishlists = @wishlists.where(user_id: params[:user_id]) if params[:user_id].present?

>     @wishlists = @wishlists.includes(:wishlist_products).where(wishlist_products: {product_id: params[:product_id]}) if params[:product_id].present?

Reviewer 3: ```suggestion
    @wishlists = @wishlists.includes(:wishlist_products).where(wishlist_products: { product_id: params[:product_id] }) if params[:product_id].present?
```

calling
[`all`](https://api.rubyonrails.org/classes/ActiveRecord/Scoping/Named/ClassMethods.html#method-i-all)
which returns an `ActiveRecord::Relation` for all the records including any
conditions we may have already applied. Then we apply the filters and return the
results

## Adding Subscribers To Admin

While we're here, we should also add the ability to view Subscribers in the
admin too. This is helpful to know how many people are waiting for a product to
go back in stock.

### Subscriber Views

First, we'll add the subscribers route to the `store` namespace:

```ruby#6
  # Admins Only
  namespace :store do
    resources :products
    resources :users
    resources :wishlists
    resources :subscribers

    root to: redirect("/store/products")
  end
```

Then, let's create the controller at
`app/controllers/store/subscribers_controller.rb`:

```ruby
class Store::SubscribersController < Store::BaseController
  before_action :set_subscriber, except: [ :index ]

  def index
    @subscribers = Subscriber.includes(:product).filter_by(params)
  end

  def show
  end

  def destroy
    @subscriber.destroy

>     redirect_to store_subscribers_path, notice: "Subscriber has been removed."

Reviewer 3: ```suggestion
    redirect_to store_subscribers_path, notice: "Subscriber has been removed.", status: :see_other
```

    <%= button_to "Delete", [ :store, @product ], method: :delete, data: { turbo_confirm: "Are you sure?" } %>
  </section>
</section>

<section>
  <%= link_to pluralize(@product.wishlists_count, "wishlist"), store_wishlists_path(product_id: @product) %>
  <%= link_to pluralize(@product.subscribers.count, "subscriber"), store_subscribers_path(product_id: @product) %>
</section>
```

## Testing Wishlists

Let's write some tests for the functionality we just built.

### Adding Fixtures

First, we need to update the fixtures in `test/fixtures/wishlist_products.yml`
so they refer to the product fixtures we have defined:

```yaml
one:
  product: tshirt
  wishlist: one

two:
  product: tshirt
  wishlist: two
```

Let's also add another product fixture in `test/fixtures/products.yml` to test
with:

```yaml#5-7
tshirt:
  name: T-Shirt
  inventory_count: 15

shoes:
  name: shoes
  inventory_count: 0
```

### Testing `to_param`

Our Wishlist model's friendly URLs are an important feature, so let's start by
writing a test for it.

Open `test/models/wishlist_test.rb` and add this test:

```ruby
require "test_helper"

class WishlistTest < ActiveSupport::TestCase

>   test "has friendly urls" do

Reviewer 2: IDK how I feel about this test. This seems mostly to be testing Rails code (which we should assume is tested). Maybe we could test that we can find it? Or maybe no test for this at all, since the integration specs will catch it.

Author: Agreed, I'll swap this with a test for `filter_by` since that is something actually worth unit testing.
